### PR TITLE
Improvements to triplanar blending

### PR DESCRIPTION
**Pull request by crydalch**
Fixes triplanar node's blending, so the corner's are not bright. Also adds a blend control for users to adjust the blending between maps. The default blend value of 1.0 matches the original behavior (apart from the brightness bug). Here is a comparison of the old triplanar (left), with the new (right):

mtlxtriplanar_orig_compare

Here are some different blend values, showing sharper or smoother blending of the maps:

mtlxtriplaner_blends

Math worked out mostly by @BrianSharpe.

### CODE REVIEW

**Comment by jstone-lucasfilm** [comment]
@crydalch @BrianSharpe I like the overall direction of this proposed change, but I suspect there may be an error in the math, as both MaterialXView and MaterialXGraphEditor render `triplanarprojection` nodes as uniform black when running with this pull request.

Here's the example MaterialX file that I'm testing, which renders as expected in the main branch, but renders as uniform black with this pull request.

[triplanar_example.zip](https://github.com/AcademySoftwareFoundation/MaterialX/files/11857858/triplanar_example.zip)

File: libraries/stdlib/stdlib_ng.mtlx [hunk 1]
```diff
@@ -265,38 +265,52 @@
...
+    </clamp>
+    <divide name="N_oneOverBlend" type="float">
+      <input name="in1" type="float" value="1.0" />
+      <input name="in2" type="float" nodename="N_clampForPrecision" />
+    </divide>
+    <power name="N_blendPower" type="vector3">
+      <input name="in1" type="vector3" nodename="N_normalizeWeights" />
+      <input name="in2" type="float" nodename="N_oneOverBlend" />
+    </power>
+    <dotproduct name="N_dotBlendedN" type="float">
```
Comment (jstone-lucasfilm) [inline libraries/stdlib/stdlib_ng.mtlx:292]: @crydalch @BrianSharpe I haven't yet dived deeply into the math of this node, but one issue I notice is that the `N_dotBlendedN` is computed but never used in the graph,  Could this perhaps be a clue as to what needs fixing?

**Comment by BrianSharpe** [comment]
Here is a fix for the color3 version
`
  <nodegraph name="NG_triplanarprojection_color3" nodedef="ND_triplanarprojection_color3">
    <extract name="N_extX_color3" type="float">
      <input name="in" type="vector3" interfacename="position" />
      <input name="index" type="integer" value="0" />
    </extract>
    <extract name="N_extY_color3" type="float">
      <input name="in" type="vector3" interfacename="position" />
      <input name="index" type="integer" value="1" />
    </extract>
    <extract name="N_extZ_color3" type="float">
      <input name="in" type="vector3" interfacename="position" />
      <input name="index" type="integer" value="2" />
    </extract>
    <combine2 name="N_vecYZ_color3" type="vector2">
      <input name="in1" type="float" nodename="N_extY_color3" />
      <input name="in2" type="float" nodename="N_extZ_color3" />
    </combine2>
    <combine2 name="N_vecXZ_color3" type="vector2">
      <input name="in1" type="float" nodename="N_extX_color3" />
      <input name="in2" type="float" nodename="N_extZ_color3" />
    </combine2>
    <combine2 name="N_vecXY_color3" type="vector2">
      <input name="in1" type="float" nodename="N_extX_color3" />
      <input name="in2" type="float" nodename="N_extY_color3" />
    </combine2>
    <image name="N_imgX_color3" type="color3">
      <input name="file" type="filename" interfacename="filex" />
      <input name="layer" type="string" interfacename="layerx" />
      <input name="default" type="color3" interfacename="default" />
      <input name="texcoord" type="vector2" nodename="N_vecYZ_color3" />
      <input name="uaddressmode" type="string" value="periodic" />
      <input name="vaddressmode" type="string" value="periodic" />
      <input name="filtertype" type="string" interfacename="filtertype" />
      <input name="framerange" type="string" interfacename="framerange" />
      <input name="frameoffset" type="integer" interfacename="frameoffset" />
      <input name="frameendaction" type="string" interfacename="frameendaction" />
    </image>
    <image name="N_imgY_color3" type="color3">
      <input name="file" type="filename" interfacename="filey" />
      <input name="layer" type="string" interfacename="layery" />
      <input name="default" type="color3" interfacename="default" />
      <input name="texcoord" type="vector2" nodename="N_vecXZ_color3" />
      <input name="uaddressmode" type="string" value="periodic" />
      <input name="vaddressmode" type="string" value="periodic" />
      <input name="filtertype" type="string" interfacename="filtertype" />
      <input name="framerange" type="string" interfacename="framerange" />
      <input name="frameoffset" type="integer" interfacename="frameoffset" />
      <input name="frameendaction" type="string" interfacename="frameendaction" />
    </image>
    <image name="N_imgZ_color3" type="color3">
      <input name="file" type="filename" interfacename="filez" />
      <input name="layer" type="string" interfacename="layerz" />
      <input name="default" type="color3" interfacename="default" />
      <input name="texcoord" type="vector2" nodename="N_vecXY_color3" />
      <input name="uaddressmode" type="string" value="periodic" />
      <input name="vaddressmode" type="string" value="periodic" />
      <input name="filtertype" type="string" interfacename="filtertype" />
      <input name="framerange" type="string" interfacename="framerange" />
      <input name="frameoffset" type="integer" interfacename="frameoffset" />
      <input name="frameendaction" type="string" interfacename="frameendaction" />
    </image>
    <normalize name="N_norm_vector3" type="vector3">
      <input name="in" type="vector3" interfacename="normal" />
    </normalize>
    <absval name="N_absN" type="vector3">
      <input name="in" type="vector3" nodename="N_norm_vector3" />
    </absval>
    <dotproduct name="N_dotN" type="float">
      <input name="in1" type="vector3" nodename="N_absN" />
      <input name="in2" type="vector3" value="1.0, 1.0, 1.0" />
    </dotproduct>
    <divide name="N_normalizeWeights" type="vector3">
      <input name="in1" type="vector3" nodename="N_absN" />
      <input name="in2" type="float" nodename="N_dotN" />
    </divide>
    <clamp name="N_clampForPrecision" type="float">
      <input name="in" type="float" interfacename="blend" />
      <input name="low" type="float" value="0.03" />
      <input name="high" type="float" value="1.0" />
    </clamp>
    <divide name="N_oneOverBlend" type="float">
      <input name="in1" type="float" value="1.0" />
      <input name="in2" type="float" nodename="N_clampForPrecision" />
    </divide>
    <power name="N_blendPower" type="vector3">
      <input name="in1" type="vector3" nodename="N_normalizeWeights" />
      <input name="in2" type="float" nodename="N_oneOverBlend" />
    </power>
    <dotproduct name="N_dotBlendedN" type="float">
      <input name="in1" type="vector3" nodename="N_blendPower" />
      <input name="in2" type="vector3" value="1.0, 1.0, 1.0" />
    </dotproduct> 
    <divide name="N_normalizeBlendedWeights" type="vector3">
      <input name="in1" type="vector3" nodename="N_blendPower" />
      <input name="in2" type="float" nodename="N_dotBlendedN" />
    </divide>
    <separate3 name="N_separateWeights" type="multioutput">
      <input name="in" type="vector3" nodename="N_normalizeBlendedWeights" />
    </separate3>
    <multiply name="N_nX_color3" type="color3">
      <input name="in1" type="color3" nodename="N_imgX_color3" />
      <input name="in2" type="float" nodename="N_separateWeights" output="outx" />
    </multiply>
    <multiply name="N_nY_color3" type="color3">
      <input name="in1" type="color3" nodename="N_imgY_color3" />
      <input name="in2" type="float" nodename="N_separateWeights" output="outy" />
    </multiply>
    <multiply name="N_nZ_color3" type="color3">
      <input name="in1" type="color3" nodename="N_imgZ_color3" />
      <input name="in2" type="float" nodename="N_separateWeights" output="outz" />
    </multiply>
    <add name="N_add1_color3" type="color3">
      <input name="in1" type="color3" nodename="N_nX_color3" />
      <input name="in2" type="color3" nodename="N_nY_color3" />
    </add>
    <add name="N_add2_color3" type="color3">
      <input name="in1" type="color3" nodename="N_add1_color3" />
      <input name="in2" type="color3" nodename="N_nZ_color3" />
    </add>
    <output name="out" type="color3" nodename="N_add2_color3" />
  </nodegraph>
`

**Comment by BrianSharpe** [comment]
Here is a fix for the color3 version

[NG_triplanarprojection_color3.txt](https://github.com/AcademySoftwareFoundation/MaterialX/files/11861547/NG_triplanarprojection_color3.txt)

**Comment by BrianSharpe** [comment]
Also, I think the blend parameter should have a UI min/max on there
`
<input name="blend" type="float" value="1.0" uimin="0.0" uimax="1.0" />
`

**Comment by crydalch** [comment]
Thanks @jstone-lucasfilm and @BrianSharpe, I've committed a fix that seems to have things working again, and adds the UI min/max for the _blend_ parm.

triplanar_blend_mtlxgrapheditor

**Review by jstone-lucasfilm — APPROVED** [review]
Thanks for this great work, @BrianSharpe and @crydalch, and I believe it's fine to allow this change in brightness as it's effectively a fix to a bug in the original node graph.

**Comment by jomaro110** [comment]
Don't know if I can comment here but this node is very basic and needs more settings.

It's missing the alpha (opacity) values for each image (X, Y, Z) and scale & rotation as well.
And it needs a projection type: (Object, World, Reference).

Thanks

**Comment by jstone-lucasfilm** [comment]
@jomaro110 This would be a great topic to bring up on the MaterialX channel of the ASWF Slack, where developers from SideFX, Autodesk, and other teams can weigh in with their thoughts.

Here is the top-level page on how to join the ASWF Slack, and you'll find the dedicated MaterialX channel there:

https://www.aswf.io/get-involved/

**Pull request merged**